source-facebook-marketing-native/source_facebook_marketing_native/client.py

```diff
@@ -220,6 +221,59 @@ def __init__(
         self.log = log
         self.permission_manager = PermissionManager(http, BASE_URL, log)
 
+    def _make_should_retry(
+        self,
+        params: FacebookRequestParams | None = None,
+    ) -> ShouldRetryProtocol:
+        """
+        Returns a callback that matches ShouldRetryProtocol and can:
+        - Retry on 500 errors up to MAX_500_ERROR_RETRY_ATTEMPTS
+        - Reduce page size when the error message indicates too much data was requested
+        """
+
+        def _should_retry(
+            status: int,
+            headers: Headers,
+            body: bytes,
+            attempt: int,
+        ) -> bool:
+            if status != 500:
+                return False
+
+            if attempt >= MAX_500_ERROR_RETRY_ATTEMPTS:
+                return False
+
+            if params is None:
+                # No params to adjust, just retry
+                return False
+
+            msg = ""
+            try:
+                payload = json.loads(body.decode("utf-8"))
+                msg = (
+                    payload.get("error", {}).get("message")
+                    or payload.get("message")
+                    or ""
+                )
+            except Exception:
+                msg = body.decode("utf-8", errors="ignore")
+
+            should_reduce = params.limit > MIN_PAGE_SIZE and (
+                "reduce" in msg.lower() or "amount of data" in msg.lower()
+            )
```

Comment:
The actual message I have noticed in Production is `Please reduce the amount of data you're asking for, then retry your request` , but decided to match on bits oft he string that still indicate we are fetching too much data in case the exact message changes.

The response also has an error code of `1`, but according to [Facebook](https://developers.facebook.com/docs/graph-api/guides/error-handling/) that is a very generic and possible transient error. So matching on the message made more sense in this case so we know when we should reduce the page size, simply retry, or stop retrying early.